src/com/ms/silverking/thread/lwt/LWTPoolImpl.java

```diff
@@ -316,70 +319,108 @@ public void addWork(BaseWorker worker, Object item, int priority) {
+      }
     }
-    
-    static void dumpWorkUnitStats(long[] workUnitStats, String label) {
-        long    cumulative;
-        
-        System.err.println(label);
-        cumulative = 0;
-        for (int i = 0; i < workUnitStats.length; i++) {
-            cumulative += workUnitStats[i] * i;
-            System.err.printf("%d\t%d\t%d\t%d\n", i, workUnitStats[i], workUnitStats[i] * i, cumulative);                
-        }
+    if (maxWorkUnit > 0) {
+      workUnitStats = new long[maxWorkUnit + 1];
+      for (LWTThread thread : threads) {
+        thread.gatherStats(workUnitStats);
+      }
+      dumpWorkUnitStats(workUnitStats, "LWTPool work unit stats: " + name);
     }
-    
-    private void dumpStats() {
-        System.err.println(getName() +"\tdumpStats");
-        System.err.println("\nactiveThreads");
-        dumpStats(activeThreads);
-        System.err.println("\nidleThreads");
-        dumpStats(idleThreads);
+  }
+
+  static void dumpWorkUnitStats(long[] workUnitStats, String label) {
+    long cumulative;
+
+    System.err.println(label);
+    cumulative = 0;
+    for (int i = 0; i < workUnitStats.length; i++) {
+      cumulative += workUnitStats[i] * i;
+      System.err.printf("%d\t%d\t%d\t%d\n", i, workUnitStats[i], workUnitStats[i] * i, cumulative);
     }
+  }
 
-    @Override
-    public void dumpStatsOnShutdown() {
-        synchronized (this) {
-            if (!dumpStatsOnShutdown) {
-                dumpStatsOnShutdown = true;
-                Runtime.getRuntime().addShutdownHook(new StatDumper());
-            }
-        }
+  private void dumpStats() {
+    System.err.println(getName() + "\tdumpStats");
+    System.err.println("\nactiveThreads");
+    dumpStats(activeThreads);
+    System.err.println("\nidleThreads");
+    dumpStats(idleThreads);
+  }
+
+  @Override
+  public void dumpStatsOnShutdown() {
+    synchronized (this) {
+      if (!dumpStatsOnShutdown) {
+        dumpStatsOnShutdown = true;
+        Runtime.getRuntime().addShutdownHook(new StatDumper());
+      }
     }
-    
-    private class StatDumper extends Thread {
-        public void run() {
-            dumpStats();
+  }
+
+  private class StatDumper extends Thread {
+    public void run() {
+      dumpStats();
+    }
+  }
+
+  @Override
+  public void debug() {
+    System.out.println(name + ":" + commonQueue.size());
+  }
+
+  @Override
+  public LWTPoolLoadStats getLoad() {
+    return loadStats;
+  }
+
+  @Override
+  public void stop() {
+    synchronized (this) {
+      controller.stop();
+      for (LWTThread thread : idleThreads) {
+        thread.lwtStop();
+        thread.setActive();
+      }
+      for (LWTThread thread : activeThreads) {
+        thread.lwtStop();
+      }
+      try {
+        NoopWork noopWork = new NoopWork(new NoopWorker());
+        for (int i = 0; i < workUnit * (activeThreads.size() + idleThreads.size()); i++) {
+          commonQueue.put(noopWork);
         }
+      } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)